### PR TITLE
[Bug] Preserve Adam first-step state during optimizer initialization

### PR DESCRIPTION
**Pull request by taking-lying-flat**
## Summary

- Reset Adam and AdamW step counters and moments after the zero-gradient, zero-LR state-materialization step.
- Preserve the allocated state tensors and their dtype/device while keeping the first real optimizer update at step 1.
- Keep this change production-only, without committing test-file changes.

## Why

`init_optim_state()` materializes lazy optimizer state by calling `optim.step()` with zero gradients and a temporary learning rate of zero. This leaves parameters unchanged, but Adam and AdamW still advance `state["step"]` from 0 to 1. Adam with coupled weight decay can also populate non-zero moments during that dummy step.

TorchFT calls `init_optim_state()` for every optimizer during construction, before any real training update. As a result, the first real Adam/AdamW update uses step-2 bias correction and diverges from an unwrapped optimizer from the first update. Calling `state_dict()` before training has the same counter side effect.

Resetting the newly materialized Adam state fixes the optimizer math without changing the generic materialization flow or state tensor layout.

## Testing

- `python -m pytest tests/unit_tests/cpu/test_state_dict_keys.py tests/unit_tests/cpu/test_optimizer_param_groups.py -q` (28 passed)
- Changed-file lint, formatting, docs, spelling, link checks: passed
- Pyrefly on `torchtitan/components/optimizer/utils.py`: 0 errors

### CODE REVIEW

**Review by fegin — APPROVED** [review]
Thanks for the fix
